src/lib/HST/Effect/Fresh.hs

```diff
@@ -57,16 +59,23 @@ freshVarPat prefix = S.PVar S.NoSrcSpan <$> freshName prefix
 -------------------------------------------------------------------------------
 -- | Interprets a computation that needs fresh variables by generating
 --   identifiers of the form @<prefix><N>@.
-runFresh :: Sem (Fresh ': r) a -> Sem r a
-runFresh = evalState Map.empty . freshToState
+runFresh :: Set String -> Sem (Fresh ': r) a -> Sem r a
+runFresh usedIdentifiers = evalState Map.empty . freshToState
  where
   -- | Reinterprets 'Fresh' in terms of 'State'.
   freshToState :: Sem (Fresh ': r) a -> Sem (State (Map String Int) ': r) a
   freshToState = reinterpret \case
     FreshIdent prefix -> do
       nextId <- gets $ Map.findWithDefault (0 :: Int) prefix
-      modify $ Map.insert prefix (nextId + 1)
-      return (prefix ++ show nextId)
+      let newId = while (\n -> (prefix ++ show n) `Set.member` usedIdentifiers)
+            (+ 1) nextId
+      modify $ Map.insert prefix (newId + 1)
+      return (prefix ++ show newId)
+
+   -- Applies a given function to a given value as long as the condition holds
```

Comment:
I think this comment is the reason why the Haddock documentation build currently fails.
```suggestion
  -- | Applies the given function repeatedly to the given value as long as the
  --   condition holds.
```